x/stake/endblock.go

```diff
@@ -252,3 +266,84 @@ func handleMatureUnbondingDelegations(k keeper.Keeper, ctx sdk.Context) ([]types
 
 	return completed, events
 }
+
+const (
+	maxProcessedRefundCount = 50
+)
+
+func handleRefundCrossStake(ctx sdk.Context, k keeper.Keeper) sdk.Events {
+	iterator := k.IteratorAllDelegations(ctx)
+	defer iterator.Close()
+	var refundEvents sdk.Events
+	count := 0
+	bcValidators := k.GetAllValidators(ctx)
+	bcValidatorsMap := make(map[string]types.Validator, len(bcValidators))
+	for _, v := range bcValidators {
+		bcValidatorsMap[v.OperatorAddr.String()] = v
+	}
+	for ; iterator.Valid(); iterator.Next() {
+		delegation := types.MustUnmarshalDelegation(k.CDC(), iterator.Key(), iterator.Value())
+		if delegation.CrossStake {
+			result := handleMsgSideChainUndelegate(ctx, types.MsgSideChainUndelegate{
+				DelegatorAddr: delegation.DelegatorAddr,
+				ValidatorAddr: delegation.ValidatorAddr,
+				Amount:        sdk.NewCoin(k.BondDenom(ctx), delegation.GetShares().RawInt()),
+				SideChainId:   k.ScKeeper.BscSideChainId(ctx),
+			}, k)
+			refundEvents = refundEvents.AppendEvents(result.Events)
+		} else {
+			if _, exist := bcValidatorsMap[delegation.DelegatorAddr.String()]; !exist {
+				result := handleMsgUndelegate(ctx, types.MsgUndelegate{
+					DelegatorAddr: delegation.DelegatorAddr,
+					ValidatorAddr: delegation.ValidatorAddr,
+					Amount:        sdk.NewCoin(k.BondDenom(ctx), delegation.GetShares().RawInt()),
+				}, k)
+				refundEvents = refundEvents.AppendEvents(result.Events)
+			}
+		}
+		count++
+		if count >= maxProcessedRefundCount {
+			break
+		}
+	}
+	return refundEvents
+}
+
+func endRefundBlock(ctx sdk.Context, k keeper.Keeper) (validatorUpdates []abci.ValidatorUpdate, completedUbds []types.UnbondingDelegation, events sdk.Events) {
```

Comment:
I don't understand why we need `endRefundBlock` function, and the logic inside this func seems duplicated.